distribution/lib/Standard/Base/0.0.0-dev/src/Internal/Time/Format/Parser.enso

```diff
@@ -0,0 +1,270 @@
+            Format_Token.Optional_Section_End -> Nothing
+            other_token ->
+                parsed_node = self.parse_common_token other_token
+                result_builder.append parsed_node
+                @Tail_Call go Nothing
+        go Nothing
+        result_builder.to_vector
+
+    ## PRIVATE
+    parse_common_token self token = case token of
+        Format_Token.Literal text ->
+            Common_Nodes.Literal text
+        Format_Token.Curly_Section inner_text ->
+            Panic.throw (Date_Time_Format_Parse_Error.Error "Unexpected section in curly braces: {"+inner_text+"}. Currently, the curly braces are only used to parametrize the two-digit year with the max year, e.g. `yy{2049}`. If you want to include a curly brace as literal, escape it with single quotes like '{"+inner_text+"}'")
+        Format_Token.Pattern character count ->
+            date_pattern = case self.mode of
+                Parser_Mode.Simple -> self.parse_simple_date_pattern character count
+                Parser_Mode.ISO_Week_Year -> self.parse_iso_week_year_pattern character count
+            any_pattern = date_pattern.if_nothing <|
+                self.parse_time_or_timezone_pattern character count
+            any_pattern.if_nothing <|
+                self.fail_invalid_pattern character count
+            any_pattern
+        _ -> Panic.throw (Illegal_State.Error "Unexpected (here) token type: "+token.to_text)
+
+    fail_invalid_pattern self character count extra_message="" =
+        Panic.throw (Date_Time_Format_Parse_Error.Error "The pattern "+(character*count)+" is not a valid pattern for the "+self.mode.pattern_format_name+" format."+extra_message)
+
+    ## PRIVATE
+    consume_token self =
+        current_position = self.position.get
+        self.position.put current_position+1
+        self.tokens.at current_position
+
+    ## PRIVATE
+    consume_curly_parameter_if_exists self =
+        current_position = self.position.get
+        case self.tokens.at current_position of
+            Format_Token.Curly_Section inner_text ->
+                self.position.put current_position+1
+                inner_text
+            _ ->
+                # If no Curly_Section is set, do not advance the pointer and return Nothing.
+                Nothing
+
+    ## PRIVATE
+    resolve_year_representation self count = case count of
+        2 ->
+            max_year = case self.consume_curly_parameter_if_exists of
+                Nothing -> default_max_year
+                parameter_text ->
+                    Integer.parse parameter_text . catch Number_Parse_Error _->
+                        Panic.throw (Date_Time_Format_Parse_Error.Error "The curly braces setting the maximum year for `yy` must be an integer, but got: {"+parameter_text+"}.")
+            Two_Digit_Year_Representation.Value max_year=max_year
+        _ ->
+            Numeric_Representation.Value count
```

Comment:
We should allow 1 to have the same as 2-digit years. I.e. `y{2029}` would mean 4 --> 2004.

Likewise, I'd probably error if >4 but see .Net allows 5. I'm not sure that 5-digit years aren't an error!

Reply:
I thought that single `y` does not stand for single-digit years, but just the year using the minimal amount of digits.

Also, how does one tell if parsing `y.M.d` `4.5.6` should parse to `2004-05-06` or `4-05-06`? While much rarer use cases, one may need to parse ancient dates.

Reply:
This mode behaves weirdly in Java:
```
jshell> var fmt = new DateTimeFormatterBuilder() . appendValueReduced(ChronoField.YEAR, 1, 5, 2000) . toFormatter()
fmt ==> ReducedValue(Year,1,5,2000)

jshell> fmt.parse("4")
$18 ==> {Year=2004},ISO

jshell> fmt.parse("24")
$19 ==> {Year=24},ISO
```

Reply:
I'd rather not go there.

Java's single `y` actually parses like this:
```
jshell> DateTimeFormatter.ofPattern("y").parse("4")
$20 ==> {Year=4},ISO

jshell> DateTimeFormatter.ofPattern("y").parse("24")
$21 ==> {Year=24},ISO

jshell> DateTimeFormatter.ofPattern("y").parse("2024")
$22 ==> {Year=2024},ISO
```

I don't think going away from the Java convention here would be good.

Reply:
my understanding was `y` was equivalent to `yy` but with optional leading `0`.
If not working in Java then yes lets just error (its a stupid format :))